# Exercise 2.A Fundamentals of Pandas

In [2]:
import numpy as np
import pandas as pd

## series

In [3]:
# Four plain objects
aa = [1, 2, 3, 4]
ab = (5, 6, 7, 8)
ac = {9, 10, 11, 12}
ad = {'first': 13, 'second': 14, 'third': 15, 'fourth': 16}

In [5]:
# Convert each to a Pandas series
ser1 = pd.Series(aa)
ser2 = pd.Series(ab)
ser3 = pd.Series(list(ac))
ser4 = pd.Series(ad)

print ('ser1:\n', ser1)
print('\nser2:\n', ser2)
print('\nser3\n', ser3)
print('\nser4\n', ser4)

ser1:
 0    1
1    2
2    3
3    4
dtype: int64

ser2:
 0    5
1    6
2    7
3    8
dtype: int64

ser3
 0     9
1    10
2    11
3    12
dtype: int64

ser4
 first     13
second    14
third     15
fourth    16
dtype: int64


In [6]:
# Creating a series from literal values

ser5 = pd.Series(17, 18, 19, 20)

# TypeError - Series() got multiple positional arguments,
# you must pass values as a list, can't pass it as seperate arguments. 

TypeError: Index(...) must be called with a collection of some kind, 18 was passed

## DataFrames from Series

In [7]:
# Creating df1 from ser1

df1 = pd.DataFrame(ser1)
df1

,0
0,1
1,2
2,3
3,4


In [8]:
# Adding ser2 and ser3 as new columns using insert() function

df1.insert(1, 'col2', ser2)
df1.insert(2, 'col3', ser3)
df1

,0,col2,col3
0,1,5,9
1,2,6,10
2,3,7,11
3,4,8,12


In [9]:
# Add ser4 as a third column (new)

df1['col4'] = ser4

df1
# ser4 was created from a dictionary so it carries named index labels
# first, second, third, and fourth.... instead of numeric ones like 0,1,2,3..
# Pandas can't align those labels to the existing numeric index.
# so therefore column 4 fills with NaN across all rows.

,0,col2,col3,col4
0,1,5,9,NaN
1,2,6,10,NaN
2,3,7,11,NaN
3,4,8,12,NaN


In [10]:
# Add a new row at index 4 using loc

df1.loc[4] = [5, 9, 12, 0]
df1

,0,col2,col3,col4
0,1,5,9,NaN
1,2,6,10,NaN
2,3,7,11,NaN
3,4,8,12,NaN
4,5,9,12,0.0


## Reading Data into a DataFrame

In [11]:
df2 = pd.read_csv('sakila_example.csv')
df2

,title,description,category,language,actor_name
0,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,PENELOPE GUINESS
1,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,CHRISTIAN GABLE
2,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,LUCILLE TRACY
3,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,SANDRA PECK
4,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,JOHNNY CAGE
...,...,...,...,...,...
257,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,GARY PENN
258,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,CHRISTOPHER WEST
259,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,DUSTIN TAUTOU
260,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,ANGELA HUDSON


In [12]:
# The top 5 records

df2.head(5)

,title,description,category,language,actor_name
0,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,PENELOPE GUINESS
1,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,CHRISTIAN GABLE
2,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,LUCILLE TRACY
3,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,SANDRA PECK
4,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,JOHNNY CAGE


In [13]:
# The last 10 records

df2.tail(10)

# Should be rows 252-261

,title,description,category,language,actor_name
252,ATTACKS HATE,A Fast-Paced Panorama of a Technical Writer An...,Sci-Fi,English,MILLA KEITEL
253,ATTACKS HATE,A Fast-Paced Panorama of a Technical Writer An...,Sci-Fi,English,GROUCHO DUNST
254,ATTACKS HATE,A Fast-Paced Panorama of a Technical Writer An...,Sci-Fi,English,BURT TEMPLE
255,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,UMA WOOD
256,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,RIP WINSLET
257,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,GARY PENN
258,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,CHRISTOPHER WEST
259,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,DUSTIN TAUTOU
260,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,ANGELA HUDSON
261,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,JAMES PITT


In [15]:
# Now the summary info

df2.info()

# From info you can see the number of rows/columns, column names,
# data types, and how many non-null values are in each column

<class 'pandas.DataFrame'>
RangeIndex: 262 entries, 0 to 261
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   title        262 non-null    str  
 1   description  262 non-null    str  
 2   category     262 non-null    str  
 3   language     262 non-null    str  
 4   actor_name   262 non-null    str  
dtypes: str(5)
memory usage: 10.4 KB


In [17]:
# The common statistics

df2.describe() 

,title,description,category,language,actor_name
count,262,262,262,262,262
unique,46,46,15,1,145
top,ARABIA DOGMA,A Touching Epistle of a Madman And a Mad Cow w...,Horror,English,OPRAH KILMER
freq,12,12,55,262,8


### shows the count, mean, std, min, max and quartiles for numeric columns only. Text columns are excluded

In [18]:
# northwind_example.csv

df3 = pd.read_csv('northwind_example.csv')
df3

,OrderID,ProductName,UnitPrice,Quantity
0,10248,Queso Cabrales,14.0,12
1,10248,Singaporean Hokkien Fried Mee,9.8,10
2,10248,Mozzarella di Giovanni,34.8,5
3,10249,Manjimup Dried Apples,42.4,40
4,10249,Tofu,18.6,9
...,...,...,...,...
495,10435,Mozzarella di Giovanni,27.8,10
496,10436,Gnocchi di nonna Alice,30.4,40
497,10436,Wimmers gute Semmelkndel,26.6,30
498,10436,Rhnbru Klosterbier,6.2,24


In [19]:
# common statistics

df3.describe()

,OrderID,UnitPrice,Quantity
count,500.000000,500.000000,500.000000
mean,10341.364000,23.371400,24.330000
std,54.442663,27.334436,18.069702
min,10248.000000,2.000000,1.000000
25%,10294.000000,10.400000,10.000000
50%,10341.000000,15.800000,20.000000
75%,10389.000000,27.800000,32.000000
max,10436.000000,210.800000,120.000000


## Insights and how it differs from sakila

- North wind has 500 rows and 4 columns, with a mix of numeric and text data. Min, max, and quartiles... unlike Sakila where all columns were text, the Unitprice ranges from '$2.00' to '$210.80' with a mean of -$23.37, suggesting most products are low-to-mid priced with a few expensive items pulling the max up.
- Quantity ordered averages about 24 units per line item, ranging from 1 to 120.
- OrderID should not be interpreted as a meaningful number, its mean and std are shown by describe() but they have no real value since its just an ID.
- Sakila's describe() only showed count, unique, top, and freq because every column was text. Northwind's describe() shows full numeric stats because it has actual quantitive columns worth mearsuing like Unitprice and Quantity.